aten/src/ATen/native/cuda/TensorModeKernel.cu

```diff
@@ -0,0 +1,322 @@
+#include <ATen/Context.h>
+#include <ATen/Dispatch.h>
+#include <ATen/TensorIterator.h>
+#include <ATen/native/ReduceOpsUtils.h>
+#include <ATen/native/TensorCompare.h>
+#include <c10/util/Exception.h>
+#include <thrust/device_ptr.h>
+#include <thrust/device_vector.h>
+#include <thrust/execution_policy.h>
+#include <thrust/extrema.h>
+#include <thrust/inner_product.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <ATen/cuda/detail/IndexUtils.cuh>
+#include <ATen/native/cuda/TensorModeKernel.cuh>
+#include <THC/THCThrustAllocator.cuh>
+
+#define MAX_GRID_SIZE 65535
+#define MAX_BLOCK_SIZE 1024
+
+namespace at {
+namespace native {
+
+template <typename scalar_t>
+void calculate_mode(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    std::vector<int64_t>& position,
+    thrust::device_vector<int64_t>& sort_buffer,
+    int dim) {
+  auto state = globalContext().getTHCState();
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  TORCH_INTERNAL_ASSERT(self.is_contiguous());
+
+  // Because the input is contiguous, we want to get a reference to the
+  // location of the buffer at the innermost dimension that we are going
+  // to calculate the mode for --> we do this by manually doing the stride
+  // calculations to get an offset
+  scalar_t* data = self.data_ptr<scalar_t>();
+  for (int64_t i = 0; i < position.size(); i++) {
+    data += position[i] * ensure_nonempty_stride(self, i);
+  }
+
+  int64_t ndim = ensure_nonempty_dim(self.dim());
+  int64_t n_element = ensure_nonempty_size(self, ndim - 1);
+  THCThrustAllocator thrust_allocator(state);
+
+  // Wrap input data in Thrust device vector
+  thrust::device_ptr<scalar_t> vec_ptr = thrust::device_pointer_cast(data);
+  thrust::device_vector<scalar_t> iter(vec_ptr, vec_ptr + n_element);
+
+  // Fill sort_buffer with [0, 1, 2, ... n_element - 1]
+  thrust::sequence(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      sort_buffer.begin(),
+      sort_buffer.end());
+
+  // Sort the input data. The original indices of the data are stored in
+  // sort_buffer
+  thrust::sort_by_key(
+      thrust::cuda::par(thrust_allocator).on(stream),
+      iter.begin(),
+      iter.end(),
+      sort_buffer.begin());
+
+  // Count # of unique elements via an inner product between adjacent elements.
+  // Add 1 if two neighboring element are not equal.
+  int unique = 1 +
+      thrust::inner_product(
+                   thrust::cuda::par(thrust_allocator).on(stream),
+                   iter.begin(),
+                   iter.end() - 1,
+                   iter.begin() + 1,
+                   0,
+                   thrust::plus<int>(),
+                   thrust::not_equal_to<scalar_t>());
+
+  // Count frequency of each element
+  thrust::device_vector<scalar_t> keys(unique);
```

Comment:
thrust::device_vector calls cudaMalloc under the hood, please avoid using it (you can allocate raw memory with cuda allocator and wrap it in thrust::device_ptr). 
Alternatively, look here for how to get counts of sorted unique elements without device_vectors https://github.com/pytorch/pytorch/blob/758fb94fcb8e20d41df6b055e80725e37ddb4854/aten/src/ATen/native/cuda/Unique.cu#L60-L68

Reply:
I tried to keep the same logic without using `thrust::device_vector`. Is that alright?

Reply:
Yeah, if you are allocating memory using pytorch allocator and not `thrust::device_vector` that should be fine.